barcode_library_characterizer/library_quality_control.py

```diff
@@ -124,24 +124,85 @@ def build_hamming_distance_histogram_parallel(
     return hamming_histogram
 
 
-def build_hamming_distance_multithreading(
-        csv_path: pathlib.Path) -> dict:
+def build_hamming_distance_simple(csv_path: pathlib.Path) -> dict:
     hamming_histogram = defaultdict(int)
-    
+
     with open(csv_path) as csv_file:
         reader = csv.reader(csv_file)
         skip_lines(reader, 1)
-        barcodes = [
-            row[1]
-            for row in reader ]
+        barcodes = [row[1] for row in reader]
 
-    pairs = [ (slow,fast) for slow in range(len(barcodes))
-              for fast in range(slow+1, len(barcodes)) ]
+    return build_histogram_simple(barcodes)
 
-    for slow,fast in pairs:
-        distance = hamming_distance(barcodes[fast], barcodes[slow])
-        hamming_histogram[distance] += 1
-    
-    return hamming_histogram
 
-    
+def build_histogram_simple(barcodes):
+    arr = [0] * (len(barcodes[0]) + 1)
+    c = Counter(
+        hamming_distance(barcodes[fast], barcodes[slow])
+        for slow, fast in pairs(len(barcodes))
+    )
+    for i, count in c.items():
+        arr[i] = count
+    return arr
+
+
+def build_histogram_arr(barcodes):
+    arr = [0] * (len(barcodes[0]) + 1)
+
+    for slow, fast in pairs(len(barcodes)):
+        d = hamming_distance(barcodes[fast], barcodes[slow])
+        arr[d] += 1
+    return arr
+
+
+# Trick to share state between processes, add it to a module field
+from . import shared
+
+
+# can't define as a local function because it can't be pickled to send between processes
+def update(i, j):
+    d = hamming_distance(shared.barcodes[i], shared.barcodes[j])
+    shared.A[d] += 1
+    # return d
+
+
+# https://stackoverflow.com/a/1721911
+def initProcess(A, barcodes):
+    shared.A = A
+    shared.barcodes = barcodes
+
+
+def build_histogram_multiprocess(barcodes):
+    arr = [0] * (len(barcodes[0]) + 1)
+    A = Array("i", arr)
+
+    with Pool(initializer=initProcess, initargs=(A, barcodes)) as p:
+        p.starmap(update, pairs(len(barcodes)))
+
+    return list(A)
+
+
+def pairs(n=10):
+    for slow in range(n):
+        for fast in range(slow + 1, n):
+            yield (slow, fast)
+
+
+def map_f(idx):
+    i, j = idx
+    arr = [0] * (len(shared.barcodes[0]) + 1)
+    d = hamming_distance(shared.barcodes[i], shared.barcodes[j])

```

Comment:
I am still wondering if it's faster to use the indexes or just pass the sequences to be compared.